// office-ui-fabric-react/lib/components/ContextualMenu/examples/ContextualMenu.Basic.Example.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var React = require("react");
var ContextualMenu_1 = require("office-ui-fabric-react/lib/ContextualMenu");
var Button_1 = require("office-ui-fabric-react/lib/Button");
require("./ContextualMenuExample.scss");
var ContextualMenuBasicExample = /** @class */ (function (_super) {
    tslib_1.__extends(ContextualMenuBasicExample, _super);
    function ContextualMenuBasicExample() {
        var _this = _super.call(this) || this;
        _this.state = {
            showCallout: false
        };
        return _this;
    }
    ContextualMenuBasicExample.prototype.render = function () {
        return (React.createElement("div", null,
            React.createElement(Button_1.DefaultButton, { id: 'ContextualMenuButton1', text: 'Click for ContextualMenu', menuProps: {
                    shouldFocusOnMount: true,
                    items: [
                        {
                            key: 'newItem',
                            name: 'New'
                        },
                        {
                            key: 'divider_1',
                            itemType: ContextualMenu_1.ContextualMenuItemType.Divider
                        },
                        {
                            key: 'rename',
                            name: 'Rename'
                        },
                        {
                            key: 'edit',
                            name: 'Edit'
                        },
                        {
                            key: 'properties',
                            name: 'Properties'
                        },
                        {
                            key: 'disabled',
                            name: 'Disabled item',
                            disabled: true
                        }
                    ]
                } })));
    };
    return ContextualMenuBasicExample;
}(React.Component));
exports.ContextualMenuBasicExample = ContextualMenuBasicExample;
