// office-ui-fabric-react/lib-es2015/components/Dialog/examples/Dialog.LargeHeader.Example.js
import * as tslib_1 from "tslib";
import * as React from 'react';
import { autobind } from '../../../Utilities';
import { Dialog, DialogType, DialogFooter } from 'office-ui-fabric-react/lib/Dialog';
import { PrimaryButton, DefaultButton } from 'office-ui-fabric-react/lib/Button';
import { ChoiceGroup } from 'office-ui-fabric-react/lib/ChoiceGroup';
var DialogLargeHeaderExample = /** @class */ (function (_super) {
    tslib_1.__extends(DialogLargeHeaderExample, _super);
    function DialogLargeHeaderExample() {
        var _this = _super.call(this) || this;
        _this.state = {
            hideDialog: true
        };
        return _this;
    }
    DialogLargeHeaderExample.prototype.render = function () {
        return (React.createElement("div", null,
            React.createElement(DefaultButton, { description: 'Opens the Sample Dialog', onClick: this._showDialog, text: 'Open Dialog' }),
            React.createElement(Dialog, { hidden: this.state.hideDialog, onDismiss: this._closeDialog, dialogContentProps: {
                    type: DialogType.largeHeader,
                    title: 'All emails together',
                    subText: 'Your Inbox has changed. No longer does it include favorites, it is a singular destination for your emails.'
                }, modalProps: {
                    isBlocking: false,
                    containerClassName: 'ms-dialogMainOverride'
                } },
                React.createElement(ChoiceGroup, { options: [
                        {
                            key: 'A',
                            text: 'Option A'
                        },
                        {
                            key: 'B',
                            text: 'Option B',
                            checked: true
                        },
                        {
                            key: 'C',
                            text: 'Option C',
                            disabled: true
                        }
                    ], onChange: this._onChoiceChanged }),
                React.createElement(DialogFooter, null,
                    React.createElement(PrimaryButton, { onClick: this._closeDialog, text: 'Save' }),
                    React.createElement(DefaultButton, { onClick: this._closeDialog, text: 'Cancel' })))));
    };
    DialogLargeHeaderExample.prototype._showDialog = function () {
        this.setState({ hideDialog: false });
    };
    DialogLargeHeaderExample.prototype._closeDialog = function () {
        this.setState({ hideDialog: true });
    };
    DialogLargeHeaderExample.prototype._onChoiceChanged = function () {
        console.log('Choice option change');
    };
    tslib_1.__decorate([
        autobind
    ], DialogLargeHeaderExample.prototype, "_showDialog", null);
    tslib_1.__decorate([
        autobind
    ], DialogLargeHeaderExample.prototype, "_closeDialog", null);
    return DialogLargeHeaderExample;
}(React.Component));
export { DialogLargeHeaderExample };
