// office-ui-fabric-react/lib-es2015/components/Dialog/examples/Dialog.Blocking.Example.js
import * as tslib_1 from "tslib";
import * as React from 'react';
import { autobind } from '../../../Utilities';
import { Dialog, DialogType, DialogFooter } from 'office-ui-fabric-react/lib/Dialog';
import { PrimaryButton, DefaultButton } from 'office-ui-fabric-react/lib/Button';
var DialogBlockingExample = /** @class */ (function (_super) {
    tslib_1.__extends(DialogBlockingExample, _super);
    function DialogBlockingExample() {
        var _this = _super.call(this) || this;
        _this.state = {
            hideDialog: true
        };
        return _this;
    }
    DialogBlockingExample.prototype.render = function () {
        return (React.createElement("div", null,
            React.createElement(DefaultButton, { description: 'Opens the Sample Dialog', onClick: this._showDialog, text: 'Open Dialog' }),
            React.createElement(Dialog, { hidden: this.state.hideDialog, onDismiss: this._closeDialog, dialogContentProps: {
                    type: DialogType.normal,
                    title: 'All emails together',
                    subText: 'Your Inbox has changed. No longer does it include favorites, it is a singular destination for your emails.'
                }, modalProps: {
                    isBlocking: true,
                    containerClassName: 'ms-dialogMainOverride'
                } },
                React.createElement(DialogFooter, null,
                    React.createElement(PrimaryButton, { onClick: this._closeDialog, text: 'Save' }),
                    React.createElement(DefaultButton, { onClick: this._closeDialog, text: 'Cancel' })))));
    };
    DialogBlockingExample.prototype._showDialog = function () {
        this.setState({ hideDialog: false });
    };
    DialogBlockingExample.prototype._closeDialog = function () {
        this.setState({ hideDialog: true });
    };
    tslib_1.__decorate([
        autobind
    ], DialogBlockingExample.prototype, "_showDialog", null);
    tslib_1.__decorate([
        autobind
    ], DialogBlockingExample.prototype, "_closeDialog", null);
    return DialogBlockingExample;
}(React.Component));
export { DialogBlockingExample };
