// @uifabric/styling/lib-amd/styles/index.js
define(["require", "exports", "tslib", "./AnimationStyles", "./DefaultPalette", "./DefaultFontStyles", "./fonts", "./getFocusStyle", "./hiddenContentStyle", "./theme", "./CommonStyles", "./GeneralStyles"], function (require, exports, tslib_1, AnimationStyles_1, DefaultPalette_1, DefaultFontStyles_1, fonts_1, getFocusStyle_1, hiddenContentStyle_1, theme_1, CommonStyles_1, GeneralStyles_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.AnimationStyles = AnimationStyles_1.AnimationStyles;
    exports.AnimationVariables = AnimationStyles_1.AnimationVariables;
    exports.DefaultPalette = DefaultPalette_1.DefaultPalette;
    exports.DefaultFontStyles = DefaultFontStyles_1.DefaultFontStyles;
    exports.registerDefaultFontFaces = DefaultFontStyles_1.registerDefaultFontFaces;
    exports.FontSizes = fonts_1.FontSizes;
    exports.FontWeights = fonts_1.FontWeights;
    exports.IconFontSizes = fonts_1.IconFontSizes;
    exports.createFontStyles = fonts_1.createFontStyles;
    exports.getFocusStyle = getFocusStyle_1.getFocusStyle;
    exports.focusClear = getFocusStyle_1.focusClear;
    exports.hiddenContentStyle = hiddenContentStyle_1.hiddenContentStyle;
    exports.ThemeSettingName = theme_1.ThemeSettingName;
    exports.getTheme = theme_1.getTheme;
    exports.loadTheme = theme_1.loadTheme;
    exports.createTheme = theme_1.createTheme;
    exports.registerOnThemeChangeCallback = theme_1.registerOnThemeChangeCallback;
    exports.removeOnThemeChangeCallback = theme_1.removeOnThemeChangeCallback;
    tslib_1.__exportStar(CommonStyles_1, exports);
    exports.normalize = GeneralStyles_1.normalize;
});
