// @microsoft/sp-build-web/lib/karma/karma.config.js
"use strict";
const path = require("path");
const Resolve_1 = require("@microsoft/sp-build-core-tasks/lib/utilities/Resolve");
const gulp_core_build_1 = require("@microsoft/gulp-core-build");
const gulp_core_build_karma_1 = require("@microsoft/gulp-core-build-karma");
const build = require("./../index");
const debugMode = process.argv.indexOf('--debug') !== -1;
const shipMode = process.argv.indexOf('--ship') !== -1 ||
    process.argv.indexOf('--production') !== -1;
const configResources = build.karma.resources; // tslint:disable-line:no-any
const bindPolyfillPath = configResources.bindPolyfillPath;
const spBuildWebRoot = path.join(__dirname, '..', '..');
const spPolyfillsRoot = Resolve_1.resolvePackage('@microsoft/sp-polyfills', spBuildWebRoot);
const spPolyfillsPath = path.join(spPolyfillsRoot, 'lib', 'index.js');
const webpackConfig = build.configureWebpack.getKarmaConfig((debugMode ? [] : [
    {
        test: /\.js$/,
        exclude: /(test|node_modules|bower_components|systemjs|requirejs)/,
        loader: configResources.istanbulInstrumenterLoaderPath
    }
]), [{
        'jsdom': 'window',
        'cheerio': 'window',
        'react/addons': true,
        'react/lib/ExecutionEnvironment': true,
        'react/lib/ReactContext': true
    }], gulp_core_build_karma_1.karma);
webpackConfig.context = gulp_core_build_1.getConfig().rootPath;
const testsJsFilePath = path.join(gulp_core_build_1.getConfig().tempFolder, 'tests.js');
module.exports = (config) => {
    const karmaConfig = {
        // base path that will be used to resolve all patterns (eg. files, exclude)
        basePath: gulp_core_build_1.getConfig().rootPath,
        // frameworks to use
        // available frameworks: https://npmjs.org/browse/keyword/karma-adapter
        frameworks: ['mocha', 'sinon-chai'],
        // list of files / patterns to load in the browser
        files: [bindPolyfillPath].concat([
            spPolyfillsPath,
            testsJsFilePath
        ]),
        // list of files to exclude
        exclude: [],
        // webpack config for bundling tests.
        webpack: webpackConfig,
        webpackMiddleware: {
            noInfo: true,
            quiet: !debugMode
        },
        // preprocess matching files before serving them to the browser
        // available preprocessors: https://npmjs.org/browse/keyword/karma-preprocessor
        preprocessors: {
            [testsJsFilePath]: ['webpack'],
            [path.join(gulp_core_build_1.getConfig().libFolder, '**/*.js')]: ['webpack'],
            [spPolyfillsPath]: ['webpack']
        },
        plugins: configResources.plugins.concat([
            require('./karma-test-result-reporter')
        ]),
        // test results reporter to use
        // possible values: 'dots', 'progress'
        // available reporters: https://npmjs.org/browse/keyword/karma-reporter
        reporters: ['test-result', 'mocha-clean', 'coverage'],
        testResultReporter: {
            outputFile: path.join(gulp_core_build_1.getConfig().tempFolder, 'testResult/result.json')
        },
        coverageReporter: {
            dir: path.join(gulp_core_build_1.getConfig().tempFolder, 'coverage'),
            reporters: [
                {
                    type: 'html',
                    subdir: 'js'
                },
                {
                    type: 'json',
                    subdir: './',
                    file: 'js-coverage.json'
                }
            ]
        },
        // web server port
        port: 9876,
        // enable / disable colors in the output (reporters and logs)
        colors: true,
        // level of logging
        // possible values: config.LOG_DISABLE || config.LOG_ERROR || config.LOG_WARN || config.LOG_INFO || config.LOG_DEBUG
        logLevel: (shipMode ? config.LOG_INFO : config.LOG_DEBUG),
        // enable / disable watching file and executing tests whenever any file changes
        autoWatch: false,
        // start these browsers
        // available browser launchers: https://npmjs.org/browse/keyword/karma-launcher
        browsers: ['PhantomJS'],
        // Continuous Integration mode
        // if true, Karma captures browsers, runs the tests and exits
        singleRun: true,
        // Increase timeouts to reduce random CI failure
        // See https://github.com/karma-runner/karma/issues/598
        captureTimeout: 10000,
        browserDisconnectTimeout: 5000,
        browserDisconnectTolerance: 5,
        browserNoActivityTimeout: 20000 // default 10000
    };
    config.set(karmaConfig);
};
