// office-ui-fabric-react/lib-es2015/components/Persona/examples/Persona.CustomRender.Example.js
import * as tslib_1 from "tslib";
import * as React from 'react';
import { autobind } from 'office-ui-fabric-react/lib/Utilities';
import { Persona, PersonaSize, PersonaPresence } from 'office-ui-fabric-react/lib/Persona';
import { Icon } from 'office-ui-fabric-react/lib/Icon';
import { TestImages } from '../../../common/TestImages';
import './PersonaExample.scss';
import * as exampleStylesImport from '../../../common/_exampleStyles.scss';
var exampleStyles = exampleStylesImport;
var examplePersona = {
    imageUrl: TestImages.personaFemale,
    imageInitials: 'AL',
    primaryText: 'Annie Lindqvist',
    secondaryText: 'Software Engineer',
    tertiaryText: 'In a meeting',
    optionalText: 'Available at 4:00pm'
};
var PersonaCustomRenderExample = /** @class */ (function (_super) {
    tslib_1.__extends(PersonaCustomRenderExample, _super);
    function PersonaCustomRenderExample() {
        return _super.call(this) || this;
    }
    PersonaCustomRenderExample.prototype.render = function () {
        return (React.createElement("div", { className: 'ms-PersonaExample' },
            React.createElement("div", { className: exampleStyles.exampleLabel }, "Custom icon in secondary text"),
            React.createElement(Persona, tslib_1.__assign({}, examplePersona, { size: PersonaSize.size72, presence: PersonaPresence.offline, onRenderSecondaryText: this._onRenderSecondaryText })),
            React.createElement("div", { className: exampleStyles.exampleLabel }, "Custom coin size = 150"),
            React.createElement(Persona, tslib_1.__assign({}, examplePersona, { coinSize: 150, presence: PersonaPresence.online, onRenderSecondaryText: this._onRenderSecondaryText }))));
    };
    PersonaCustomRenderExample.prototype._onRenderSecondaryText = function (props) {
        return (React.createElement("div", null,
            React.createElement(Icon, { iconName: 'Suitcase', className: 'ms-JobIconExample' }),
            props.secondaryText));
    };
    tslib_1.__decorate([
        autobind
    ], PersonaCustomRenderExample.prototype, "_onRenderSecondaryText", null);
    return PersonaCustomRenderExample;
}(React.Component));
export { PersonaCustomRenderExample };
