// office-ui-fabric-react/lib/components/DetailsList/examples/DetailsList.NavigatingFocus.Example.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var React = require("react");
var DetailsList_1 = require("office-ui-fabric-react/lib/DetailsList");
var Link_1 = require("office-ui-fabric-react/lib/Link");
var Utilities_1 = require("office-ui-fabric-react/lib/Utilities");
require("./DetailsListExample.scss");
var DetailsListNavigatingFocusExample = /** @class */ (function (_super) {
    tslib_1.__extends(DetailsListNavigatingFocusExample, _super);
    function DetailsListNavigatingFocusExample() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.state = {
            items: generateItems(''),
            initialFocusedIndex: undefined,
        };
        _this._columns = [
            {
                name: 'File path',
                onRender: function (item) {
                    return (React.createElement(Link_1.Link, { key: item, onClick: _this._navigate(item) }, item));
                },
            },
            {
                name: 'Size',
                onRender: function (item) { return '4 KB'; },
            }
        ];
        return _this;
    }
    DetailsListNavigatingFocusExample.prototype.render = function () {
        return (React.createElement(DetailsList_1.DetailsList, { items: this.state.items, columns: this._columns, initialFocusedIndex: this.state.initialFocusedIndex }));
    };
    DetailsListNavigatingFocusExample.prototype._navigate = function (name) {
        var _this = this;
        return function () {
            _this.setState({
                items: generateItems(name + '/'),
                initialFocusedIndex: 0,
            });
        };
    };
    tslib_1.__decorate([
        Utilities_1.autobind
    ], DetailsListNavigatingFocusExample.prototype, "_navigate", null);
    return DetailsListNavigatingFocusExample;
}(React.Component));
exports.DetailsListNavigatingFocusExample = DetailsListNavigatingFocusExample;
function generateItems(parent) {
    return Array.prototype.map.call('abcdefghi', function (name) { return parent + name; });
}
