// office-ui-fabric-react/lib-es2015/components/ContextualMenu/examples/ContextualMenu.Basic.Example.js
import * as tslib_1 from "tslib";
import * as React from 'react';
import { ContextualMenuItemType } from 'office-ui-fabric-react/lib/ContextualMenu';
import { DefaultButton } from 'office-ui-fabric-react/lib/Button';
import './ContextualMenuExample.scss';
var ContextualMenuBasicExample = /** @class */ (function (_super) {
    tslib_1.__extends(ContextualMenuBasicExample, _super);
    function ContextualMenuBasicExample() {
        var _this = _super.call(this) || this;
        _this.state = {
            showCallout: false
        };
        return _this;
    }
    ContextualMenuBasicExample.prototype.render = function () {
        return (React.createElement("div", null,
            React.createElement(DefaultButton, { id: 'ContextualMenuButton1', text: 'Click for ContextualMenu', menuProps: {
                    shouldFocusOnMount: true,
                    items: [
                        {
                            key: 'newItem',
                            name: 'New'
                        },
                        {
                            key: 'divider_1',
                            itemType: ContextualMenuItemType.Divider
                        },
                        {
                            key: 'rename',
                            name: 'Rename'
                        },
                        {
                            key: 'edit',
                            name: 'Edit'
                        },
                        {
                            key: 'properties',
                            name: 'Properties'
                        },
                        {
                            key: 'disabled',
                            name: 'Disabled item',
                            disabled: true
                        }
                    ]
                } })));
    };
    return ContextualMenuBasicExample;
}(React.Component));
export { ContextualMenuBasicExample };
