// office-ui-fabric-react/lib-amd/components/Dialog/examples/Dialog.Blocking.Example.js
define(["require", "exports", "tslib", "react", "../../../Utilities", "office-ui-fabric-react/lib/Dialog", "office-ui-fabric-react/lib/Button"], function (require, exports, tslib_1, React, Utilities_1, Dialog_1, Button_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var DialogBlockingExample = /** @class */ (function (_super) {
        tslib_1.__extends(DialogBlockingExample, _super);
        function DialogBlockingExample() {
            var _this = _super.call(this) || this;
            _this.state = {
                hideDialog: true
            };
            return _this;
        }
        DialogBlockingExample.prototype.render = function () {
            return (React.createElement("div", null,
                React.createElement(Button_1.DefaultButton, { description: 'Opens the Sample Dialog', onClick: this._showDialog, text: 'Open Dialog' }),
                React.createElement(Dialog_1.Dialog, { hidden: this.state.hideDialog, onDismiss: this._closeDialog, dialogContentProps: {
                        type: Dialog_1.DialogType.normal,
                        title: 'All emails together',
                        subText: 'Your Inbox has changed. No longer does it include favorites, it is a singular destination for your emails.'
                    }, modalProps: {
                        isBlocking: true,
                        containerClassName: 'ms-dialogMainOverride'
                    } },
                    React.createElement(Dialog_1.DialogFooter, null,
                        React.createElement(Button_1.PrimaryButton, { onClick: this._closeDialog, text: 'Save' }),
                        React.createElement(Button_1.DefaultButton, { onClick: this._closeDialog, text: 'Cancel' })))));
        };
        DialogBlockingExample.prototype._showDialog = function () {
            this.setState({ hideDialog: false });
        };
        DialogBlockingExample.prototype._closeDialog = function () {
            this.setState({ hideDialog: true });
        };
        tslib_1.__decorate([
            Utilities_1.autobind
        ], DialogBlockingExample.prototype, "_showDialog", null);
        tslib_1.__decorate([
            Utilities_1.autobind
        ], DialogBlockingExample.prototype, "_closeDialog", null);
        return DialogBlockingExample;
    }(React.Component));
    exports.DialogBlockingExample = DialogBlockingExample;
});
