// office-ui-fabric-react/lib-amd/components/Dialog/examples/Dialog.LargeHeader.Example.js
define(["require", "exports", "tslib", "react", "../../../Utilities", "office-ui-fabric-react/lib/Dialog", "office-ui-fabric-react/lib/Button", "office-ui-fabric-react/lib/ChoiceGroup"], function (require, exports, tslib_1, React, Utilities_1, Dialog_1, Button_1, ChoiceGroup_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var DialogLargeHeaderExample = /** @class */ (function (_super) {
        tslib_1.__extends(DialogLargeHeaderExample, _super);
        function DialogLargeHeaderExample() {
            var _this = _super.call(this) || this;
            _this.state = {
                hideDialog: true
            };
            return _this;
        }
        DialogLargeHeaderExample.prototype.render = function () {
            return (React.createElement("div", null,
                React.createElement(Button_1.DefaultButton, { description: 'Opens the Sample Dialog', onClick: this._showDialog, text: 'Open Dialog' }),
                React.createElement(Dialog_1.Dialog, { hidden: this.state.hideDialog, onDismiss: this._closeDialog, dialogContentProps: {
                        type: Dialog_1.DialogType.largeHeader,
                        title: 'All emails together',
                        subText: 'Your Inbox has changed. No longer does it include favorites, it is a singular destination for your emails.'
                    }, modalProps: {
                        isBlocking: false,
                        containerClassName: 'ms-dialogMainOverride'
                    } },
                    React.createElement(ChoiceGroup_1.ChoiceGroup, { options: [
                            {
                                key: 'A',
                                text: 'Option A'
                            },
                            {
                                key: 'B',
                                text: 'Option B',
                                checked: true
                            },
                            {
                                key: 'C',
                                text: 'Option C',
                                disabled: true
                            }
                        ], onChange: this._onChoiceChanged }),
                    React.createElement(Dialog_1.DialogFooter, null,
                        React.createElement(Button_1.PrimaryButton, { onClick: this._closeDialog, text: 'Save' }),
                        React.createElement(Button_1.DefaultButton, { onClick: this._closeDialog, text: 'Cancel' })))));
        };
        DialogLargeHeaderExample.prototype._showDialog = function () {
            this.setState({ hideDialog: false });
        };
        DialogLargeHeaderExample.prototype._closeDialog = function () {
            this.setState({ hideDialog: true });
        };
        DialogLargeHeaderExample.prototype._onChoiceChanged = function () {
            console.log('Choice option change');
        };
        tslib_1.__decorate([
            Utilities_1.autobind
        ], DialogLargeHeaderExample.prototype, "_showDialog", null);
        tslib_1.__decorate([
            Utilities_1.autobind
        ], DialogLargeHeaderExample.prototype, "_closeDialog", null);
        return DialogLargeHeaderExample;
    }(React.Component));
    exports.DialogLargeHeaderExample = DialogLargeHeaderExample;
});
