// office-ui-fabric-react/lib/components/Panel/examples/Panel.SmallRight.Example.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var React = require("react");
var Utilities_1 = require("office-ui-fabric-react/lib/Utilities");
var Button_1 = require("office-ui-fabric-react/lib/Button");
var ChoiceGroup_1 = require("office-ui-fabric-react/lib/ChoiceGroup");
var Panel_1 = require("office-ui-fabric-react/lib/Panel");
var PanelSmallRightExample = /** @class */ (function (_super) {
    tslib_1.__extends(PanelSmallRightExample, _super);
    function PanelSmallRightExample() {
        var _this = _super.call(this) || this;
        _this.state = {
            showPanel: false
        };
        return _this;
    }
    PanelSmallRightExample.prototype.render = function () {
        return (React.createElement("div", null,
            React.createElement(Button_1.DefaultButton, { description: 'Opens the Sample Panel', onClick: this._onShowPanel, text: 'Open Panel' }),
            React.createElement(Panel_1.Panel, { isOpen: this.state.showPanel, type: Panel_1.PanelType.smallFixedFar, onDismiss: this._onClosePanel, headerText: 'Panel - Small, right-aligned, fixed, with footer', closeButtonAriaLabel: 'Close', onRenderFooterContent: this._onRenderFooterContent },
                React.createElement(ChoiceGroup_1.ChoiceGroup, { options: [
                        {
                            key: 'A',
                            text: 'Option A'
                        },
                        {
                            key: 'B',
                            text: 'Option B',
                            checked: true
                        },
                        {
                            key: 'C',
                            text: 'Option C',
                            disabled: true
                        },
                        {
                            key: 'D',
                            text: 'Option D',
                            checked: true,
                            disabled: true
                        }
                    ], label: 'Pick one', required: true }))));
    };
    PanelSmallRightExample.prototype._onClosePanel = function () {
        this.setState({ showPanel: false });
    };
    PanelSmallRightExample.prototype._onRenderFooterContent = function () {
        return (React.createElement("div", null,
            React.createElement(Button_1.PrimaryButton, { onClick: this._onClosePanel, style: { 'marginRight': '8px' } }, "Save"),
            React.createElement(Button_1.DefaultButton, { onClick: this._onClosePanel }, "Cancel")));
    };
    PanelSmallRightExample.prototype._onShowPanel = function () {
        this.setState({ showPanel: true });
    };
    tslib_1.__decorate([
        Utilities_1.autobind
    ], PanelSmallRightExample.prototype, "_onClosePanel", null);
    tslib_1.__decorate([
        Utilities_1.autobind
    ], PanelSmallRightExample.prototype, "_onRenderFooterContent", null);
    tslib_1.__decorate([
        Utilities_1.autobind
    ], PanelSmallRightExample.prototype, "_onShowPanel", null);
    return PanelSmallRightExample;
}(React.Component));
exports.PanelSmallRightExample = PanelSmallRightExample;
