// office-ui-fabric-react/lib/components/ActivityItem/examples/ActivityItem.Compact.Example.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
/* tslint:disable:no-unused-variable */
var React = require("react");
/* tslint:enable:no-unused-variable */
var ActivityItem_1 = require("../ActivityItem");
var TestImages_1 = require("../../../common/TestImages");
var Icon_1 = require("../../../Icon");
var activityItemExamples = [
    {
        key: 1,
        activityDescription: [
            React.createElement("span", { key: 1, className: 'ms-activityItem-nameText' }, "Tahlia\tWhittle"),
            React.createElement("span", { key: 2 }, " edited this file")
        ],
        activityPersonas: [
            {
                imageUrl: TestImages_1.TestImages.personaFemale
            }
        ],
        isCompact: true
    },
    {
        key: 2,
        activityDescription: [
            React.createElement("span", { key: 1, className: 'ms-activityItem-nameText' }, "Patrick Loton"),
            React.createElement("span", { key: 2 }, " and "),
            React.createElement("span", { key: 3, className: 'ms-activityItem-nameText' }, " 6 others")
        ],
        activityPersonas: [
            {
                imageInitials: 'PT',
                primaryText: 'Robert Larsson'
            },
            {
                imageUrl: TestImages_1.TestImages.personaMale
            },
            {
                imageInitials: 'EC',
                primaryText: 'Eduarda Costa'
            }
        ],
        isCompact: true
    },
    {
        key: 3,
        activityDescription: [
            React.createElement("span", { key: 1, className: 'ms-activityItem-nameText' }, "Sabrina De Luca"),
            React.createElement("span", { key: 2 }, " added this file")
        ],
        activityIcon: React.createElement(Icon_1.Icon, { iconName: 'Add' }),
        isCompact: true
    },
    {
        key: 4,
        activityDescription: [
            React.createElement("span", { key: 1, className: 'ms-activityItem-nameText' }, "Chuan Rojumanong"),
            React.createElement("span", { key: 2 }, " shared this file")
        ],
        activityIcon: React.createElement(Icon_1.Icon, { iconName: 'Share' }),
        isCompact: true
    }
];
var ActivityItemCompactExample = /** @class */ (function (_super) {
    tslib_1.__extends(ActivityItemCompactExample, _super);
    function ActivityItemCompactExample() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ActivityItemCompactExample.prototype.render = function () {
        var activityExampleList = [];
        activityItemExamples.forEach(function (item) {
            var props = item;
            activityExampleList.push(React.createElement(ActivityItem_1.ActivityItem, tslib_1.__assign({}, props, { key: item.key, className: 'ms-activityItem-exampleRoot' })));
        });
        return (React.createElement("div", null, activityExampleList));
    };
    return ActivityItemCompactExample;
}(React.Component));
exports.ActivityItemCompactExample = ActivityItemCompactExample;
