// office-ui-fabric-react/lib-es2015/components/ComboBox/ComboBox.classNames.js
import { memoizeFunction } from '../../Utilities';
import { mergeStyles, } from '../../Styling';
export var getClassNames = memoizeFunction(function (styles, className, isOpen, disabled, required, focused, allowFreeForm, hasErrorMessage) {
    return {
        container: mergeStyles('ms-ComboBox-container', className, styles.container),
        label: mergeStyles(styles.label, disabled && styles.labelDisabled),
        root: mergeStyles('ms-ComboBox', isOpen && 'is-open', required && 'is-required', styles.root, !allowFreeForm && styles.rootDisallowFreeForm, hasErrorMessage && styles.rootError, !disabled && focused && styles.rootFocused, !disabled && {
            selectors: {
                ':hover': styles.rootHovered
            }
        }, disabled && [
            'is-disabled', styles.rootDisabled
        ]),
        input: mergeStyles('ms-ComboBox-Input', styles.input, disabled && styles.inputDisabled),
        errorMessage: mergeStyles(styles.errorMessage),
        callout: mergeStyles('ms-ComboBox-callout', styles.callout),
        optionsContainerWrapper: mergeStyles('ms-ComboBox-optionsContainerWrapper', styles.optionsContainerWrapper),
        optionsContainer: mergeStyles('ms-ComboBox-optionsContainer', styles.optionsContainer),
        header: mergeStyles('ms-ComboBox-header', styles.header),
        divider: mergeStyles('ms-ComboBox-divider', styles.divider),
    };
});
export var getComboBoxOptionClassNames = memoizeFunction(function (styles) {
    return {
        optionText: mergeStyles('ms-ComboBox-optionText', styles.optionText),
    };
});
