// office-ui-fabric-react/lib-es2015/components/Label/Label.classNames.js
import { memoizeFunction } from '../../Utilities';
import { mergeStyleSets } from '../../Styling';
var MS_HIGHCONTRAST_ACTIVE = '@media screen and (-ms-high-contrast: active)';
export var getLabelClassNames = memoizeFunction(function (theme, className, disabled, required) {
    return mergeStyleSets({
        root: ['ms-Label',
            {
                color: theme.semanticColors.bodyText,
                boxSizing: 'border-box',
                boxShadow: 'none',
                margin: 0,
                display: 'block',
                padding: '5px 0',
                wordWrap: 'break-word',
                overflowWrap: 'break-word',
            },
            disabled && {
                color: theme.semanticColors.disabledBodyText,
                selectors: (_a = {},
                    _a[MS_HIGHCONTRAST_ACTIVE] = {
                        color: 'GrayText'
                    },
                    _a)
            },
            required && {
                selectors: {
                    '::after': {
                        content: "' *'",
                        color: theme.semanticColors.errorText,
                        paddingRight: 12
                    }
                }
            },
            className]
    });
    var _a;
});
