// office-ui-fabric-react/lib-amd/components/Button/CommandBarButton/CommandBarButton.styles.js
define(["require", "exports", "../../../Styling", "../../../Utilities", "../BaseButton.styles", "../SplitButton/SplitButton.styles"], function (require, exports, Styling_1, Utilities_1, BaseButton_styles_1, SplitButton_styles_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.getStyles = Utilities_1.memoizeFunction(function (theme, customStyles, focusInset, focusColor) {
        var baseButtonStyles = BaseButton_styles_1.getStyles(theme);
        var baseSplitButtonStyles = SplitButton_styles_1.getStyles(theme);
        var commandButtonStyles = {
            root: {
                minWidth: '40px',
                backgroundColor: theme.palette.neutralLighter,
                color: theme.palette.neutralPrimary,
                padding: '0 4px'
            },
            rootHovered: {
                backgroundColor: theme.palette.neutralLight,
                color: theme.palette.neutralDark
            },
            rootPressed: {
                backgroundColor: theme.palette.neutralQuaternaryAlt,
                color: theme.palette.black
            },
            rootChecked: {
                backgroundColor: theme.palette.neutralQuaternaryAlt,
                color: theme.palette.black
            },
            rootExpanded: {
                backgroundColor: theme.palette.neutralQuaternaryAlt,
                color: theme.palette.black
            },
            rootCheckedHovered: {
                backgroundColor: theme.palette.neutralQuaternary,
                color: theme.palette.black,
            },
            label: {
                fontWeight: 'normal' // theme.fontWeights.semibold,
            },
            icon: {
                color: theme.palette.themeDarkAlt
            },
            menuIcon: {
                color: theme.palette.neutralSecondary
            }
        };
        return Styling_1.concatStyleSets(baseButtonStyles, commandButtonStyles, baseSplitButtonStyles, customStyles);
    });
});
